File: web/src/lib/protocol.ts
// TypeScript types mirroring crates/telepair-core/src/protocol.rs

// Stable error codes carried by ServerMessage.Error. Keep in sync with
// `pub mod error_codes` in crates/telepair-core/src/protocol.rs — a typo
// on either side silently degrades UX.
export const ErrorCode = {
  AUTH_FAILED: 'AUTH_FAILED',
  AUTH_TIMEOUT: 'AUTH_TIMEOUT',
  EXPECTED_JOIN: 'EXPECTED_JOIN',
  SESSION_NOT_FOUND: 'SESSION_NOT_FOUND',
  SESSION_CLOSED: 'SESSION_CLOSED',
  NOT_PARTICIPANT: 'NOT_PARTICIPANT',
  TARGET_NOT_FOUND: 'TARGET_NOT_FOUND',
  PTY_ERROR: 'PTY_ERROR',
  /**
   * Transient storage failure during the WS handshake (e.g. the
   * participants lookup errored out). Clients should treat it as
   * "retry in a moment" — it is NOT a permission or invite problem.
   * Paired with `CloseCode.TRANSIENT` on the close frame.
   */
  STORAGE_ERROR: 'STORAGE_ERROR',
  SESSION_DISABLED: 'SESSION_DISABLED',
} as const;
export type ErrorCode = (typeof ErrorCode)[keyof typeof ErrorCode];

/**
 * WebSocket close codes the gateway uses to signal "retry vs give up"
 * to the client. MUST stay in sync with `CLOSE_CODE_TERMINAL` /
 * `CLOSE_CODE_TRANSIENT` in `crates/telepair-core/src/protocol.rs` —
 * the close code is the single signal the client has to decide whether
 * to reconnect, since the preceding JSON `Error` frame may be dropped
 * if the socket tears down mid-write.
 */
export const CloseCode = {
  /**
   * Permanent refusal — auth / permission / not-found / target-missing.
   * `TelepairSocket.onclose` MUST surface an error state and NOT
   * schedule a reconnect, otherwise a revoked token would DoS the
   * gateway with a retry storm.
   */
  TERMINAL: 4001,
  /**
   * Transient failure (e.g. a one-off SQLite hiccup during the
   * handshake). The client is expected to reconnect on its own. Sits
   * in the private-use range (4000-4999) and is chosen to be visually
   * distinct from TERMINAL rather than for any IANA meaning.
   */
  TRANSIENT: 4503,
} as const;
export type CloseCode = (typeof CloseCode)[keyof typeof CloseCode];

export type Role = 'owner' | 'operator' | 'viewer';

export type InputMode = 'serialized' | 'multiplexed';

export type SessionStatus = 'active' | 'closed';

/**
 * Reason a session moved from `active` to `closed`. Mirrors the
 * `CloseReason` enum in `crates/telepair-core/src/session.rs` — the
 * server serializes it as a lowercase string and the history view
 * renders each variant as a distinct chip, so a new variant on the
 * backend MUST be added here at the same time or the UI will show
 * it as `Unknown`.
 *
 * - `owner`   – the owner clicked Close in the session page.
 * - `reaper`  – the idle reaper closed it after no participants were left.
 * - `startup` – server restart cleaned up an orphaned active row.
 * - `error`   – WS-phase launch failure (target vanished, PTY spawn failed, etc.).
 */
export type CloseReason = 'owner' | 'reaper' | 'startup' | 'error';

export interface Session {
  id: string;
  owner_id: string;
  target_name: string;
  input_mode: InputMode;
  status: SessionStatus;
  created_at: string;
  closed_at: string | null;
  /**
   * Populated only for closed rows. `null` on a session closed before
   * the v0.1.1 upgrade added the column, so the UI has to tolerate
   * `null` even on `status === 'closed'` entries.
   */
  closed_reason?: CloseReason | null;
}

export interface ParticipantInfo {
  user_id: string;
  name: string;
  role: Role;
  color: string;
}

export interface TargetInfo {
  name: string;
  display: string;
  tags: string[];
  /** "global" for targets.yaml entries; "user" for user-owned targets. */
  source: 'global' | 'user';
  /** Present only for user-owned targets. */
  id?: string;
  admin_only: boolean;
}

export interface UserTargetInfo {
  id: string;
  user_id: string;
  name: string;
  display: string;
  command: string;
  args: string[];
  env: Record<string, string>;
  tags: string[];
  created_at: string;
  updated_at: string;
}

// --- Client → Server ---
//
// Terminal input is NOT a JSON message — it is sent as a raw binary WebSocket
// frame (see TelepairSocket.sendInput). All other messages below are JSON.

export type ClientMessage =
  | { type: 'SessionJoin'; session_id: string; token: string; cols: number; rows: number }
  | { type: 'TermResize'; cols: number; rows: number }
  | { type: 'CursorMove'; x: number; y: number }
  | { type: 'ChatMessage'; text: string };

// --- Server → Client ---

export type ServerMessage =
  | { type: 'SessionState'; session: Session; participants: ParticipantInfo[]; your_role: Role; your_user_id: string }
  | { type: 'PeerJoined'; user_id: string; name: string; role: Role; color: string }
  | { type: 'PeerLeft'; user_id: string }
  | { type: 'PeerCursor'; user_id: string; x: number; y: number }
  | { type: 'PeerChat'; user_id: string; name: string; text: string; ts: string }
  | { type: 'InputDenied'; reason: InputDeniedReason }
  | { type: 'Error'; code: string; message: string };

// Reason codes mirroring `crates/telepair-core/src/protocol.rs::input_denied`.
// The server only sends these strings; any unknown value should be rendered
// as a generic "input not allowed" so a protocol upgrade doesn't silently
// swallow the notice.
export const InputDeniedReason = {
  VIEWER: 'VIEWER',
  SERIALIZED_NOT_OWNER: 'SERIALIZED_NOT_OWNER',
} as const;
export type InputDeniedReason = (typeof InputDeniedReason)[keyof typeof InputDeniedReason];

/**
 * Audit event type discriminant, mirroring `AuditEventType::as_str()`
 * in `crates/telepair-core/src/audit.rs`. The on-disk / on-wire form is
 * the dotted-lowercase string — this constant pins the exact set so the
 * frontend's `eventLabel` switch stays exhaustive under TypeScript and
 * a backend rename surfaces as a compile error instead of a rendered
 * raw string.
 */
export const AuditEventType = {
  SESSION_CREATED: 'session.created',
  SESSION_CLOSED: 'session.closed',
  PARTICIPANT_JOINED: 'participant.joined',
  INVITE_MINTED: 'invite.minted',
  INVITE_REDEEMED: 'invite.redeemed',
  INVITE_REVOKED: 'invite.revoked',
  TARGET_ACCESS_DENIED: 'target.access_denied',
  TARGET_RELOADED: 'target.reloaded',
  AUTH_LOGIN_FAILED: 'auth.login_failed',
  AUTH_REGISTER_REJECTED: 'auth.register_rejected',
  AUTH_REGISTER_COMPLETED: 'auth.register_completed',
  AUTH_VERIFY_FAILED: 'auth.verify_failed',
  AUTH_USER_ENABLED: 'auth.user_enabled',
  AUTH_USER_DISABLED: 'auth.user_disabled',
  AUTH_SESSION_ACCESS_DENIED: 'auth.session_access_denied',
} as const;
export type AuditEventType = (typeof AuditEventType)[keyof typeof AuditEventType];

// --- Helpers ---

const textEncoder = new TextEncoder();

export function encodeInput(text: string): Uint8Array {
  return textEncoder.encode(text);
}

/**
 * Returns whether the given (role, inputMode) combo may forward
 * keystrokes to the PTY. The old 1-arg signature silently let operators
 * type in serialized sessions — the server then dropped those bytes,
 * producing a dead-keyboard UX. Callers MUST pass `inputMode` so the
 * client and server enforcement agree.
 */
export function canInput(role: Role, inputMode: InputMode): boolean {
  if (role === 'owner') return true;
  if (role === 'operator') return inputMode === 'multiplexed';
  return false;
}

export interface InviteInfo {
  token: string;
  role: Role;
  max_uses: number;
  /**
   * Absolute expiry resolved server-side. `null` means the invite
   * never expires on its own (it will still die when `max_uses` runs
   * out or the session closes). The frontend should render a humanised
   * countdown when this is set so the owner can see at a glance how
   * long the link will stay live.
   */
  expires_at: string | null;
  session_id: string;
}

export interface RedeemResult {
  session_id: string;
  role: Role;
  /**
   * Freshly issued guest token when the redeem call was made without
   * (or with an invalid) bearer token. The frontend should store this
   * into auth state before navigating to the session. `null` when the
   * caller was already authenticated — they keep using their own
   * token in that case.
   */
  token: string | null;
}

/**
 * One row of the per-session audit timeline. Mirrors `AuditEvent` in
 * `crates/telepair-core/src/audit.rs`. The server returns rows newest
 * first (`ts DESC`), capped at 500.
 *
 * `event_type` is the dotted-lowercase form (`session.created`,
 * `participant.joined`, …) — the same string the CLI `--type` flag
 * accepts and the same column value stored in `audit_events.event_type`.
 * The list of variants is intentionally finite; an unknown value most
 * likely means the frontend is older than the backend it's talking to,
 * so the UI renders unknown types verbatim instead of throwing.
 *
 * `actor_id` and `actor_name` are nullable independently:
 *   - `actor_id` is null for events that happened without an
 *     authenticated identity (e.g. invite redemption by an anonymous
 *     visitor — the row is stamped with `actor_name='guest'` and a
 *     freshly minted user id, but pre-redeem rows have neither).
 *   - `actor_name` is a denormalized snapshot taken at insertion time;
 *     renaming a user later does NOT rewrite history.
 *
 * `detail` is an opaque JSON value (object, string, number, or null).
 * The shape varies per event type — see the matching variant comments
 * in `crates/telepair-core/src/audit.rs::AuditEventType`. The UI does
 * not assume a specific shape; it pretty-prints the JSON on demand and
 * extracts the few well-known keys (`role`, `reason`) lazily.
 */
export interface AuditEvent {
  id: number | null;
  ts: string;
  actor_id: string | null;
  actor_name: string | null;
  event_type: string;
  session_id: string | null;
  detail: unknown;
}

/**
 * Sanitized view of a single invite row for the owner-facing
 * management dialog. Mirrors `InviteSummary` in
 * `crates/telepair-control/src/invite_service.rs` — the backend
 * deliberately does NOT leak the raw bearer token here. The UI uses
 * `token_prefix` (first 8 chars of the sha) as a stable per-row label
 * and `token_sha256` as the DELETE path parameter when revoking.
 */
export interface InviteSummary {
  token_sha256: string;
  token_prefix: string;
  session_id: string;
  role: Role;
  max_uses: number;
  used_count: number;
  /** `max_uses - used_count`, clamped to zero. Precomputed server-side so every client renders the same number. */
  remaining_uses: number;
  expires_at: string | null;
  created_at: string | null;
}

/**
 * Env var presence marker for the admin-targets detail view. The
 * server deliberately NEVER returns the resolved value — leaking
 * `PGPASSWORD=...` through an HTTP API would widen the blast radius
 * beyond the "anyone who can write targets.yaml can exfiltrate env"
 * trust boundary we already accept. The UI renders `set: true` as a
 * filled chip and `set: false` as a hollow chip so an admin can spot
 * a missing variable at a glance.
 */
export interface AdminTargetEnvKey {
  key: string;
  set: boolean;
}

/**
 * One row returned by `GET /api/admin/targets`. Mirrors
 * `AdminTargetInfo` in `crates/telepair-gateway/src/http.rs`. Unlike
 * the public `TargetInfo`, this carries the full config (command,
 * args, shell) plus the runtime `active_sessions` count — everything
 * the admin page needs to render a complete detail card and a
 * per-target deep link into the session history view.
 */
export interface AdminTargetInfo {
  name: string;
  display: string;
  /** `virtual` (from targets.yaml) or `local` (the built-in local-shell target). */
  type: 'virtual' | 'local' | string;
  command: string | null;
  args: string[];
  shell: string | null;
  tags: string[];
  admin_only: boolean;
  env: AdminTargetEnvKey[];
  active_sessions: number;
}

/**
 * Success body returned by `POST /api/admin/targets/reload`. `path`
 * is the absolute path re-read from disk; `targets` is the number of
 * targets in the new engine. The UI surfaces both in the success
 * toast so the admin can confirm they reloaded the file they meant to.
 */
export interface ReloadTargetsResult {
  path: string;
  targets: number;
}

/**
 * Mirrors `AdminUserInfo` in `crates/telepair-gateway/src/http.rs`.
 * Returned by `GET /api/admin/users` and
 * `POST /api/admin/users/{id}/{enable,disable}`.
 */
export interface AdminUserInfo {
  id: string;
  name: string;
  email: string | null;
  is_admin: boolean;
  session_enabled: boolean;
  created_at: string;
  updated_at: string;
}

File: crates/telepair-core/src/protocol.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::permission::Role;
use crate::session::Session;

/// Stable error codes carried by `ServerMessage::Error`. The frontend
/// switches on these strings to render localized messages and decide
/// whether to force re-login — a typo on either side silently degrades
/// UX, so the Rust and TS sides MUST stay in sync via constants, not
/// scattered string literals. Mirror table lives in `web/src/lib/protocol.ts`.
pub mod error_codes {
    pub const AUTH_FAILED: &str = "AUTH_FAILED";
    pub const AUTH_TIMEOUT: &str = "AUTH_TIMEOUT";
    pub const EXPECTED_JOIN: &str = "EXPECTED_JOIN";
    pub const SESSION_NOT_FOUND: &str = "SESSION_NOT_FOUND";
    pub const SESSION_CLOSED: &str = "SESSION_CLOSED";
    pub const NOT_PARTICIPANT: &str = "NOT_PARTICIPANT";
    pub const TARGET_NOT_FOUND: &str = "TARGET_NOT_FOUND";
    pub const PTY_ERROR: &str = "PTY_ERROR";
    /// Transient storage failure during the WS handshake. Clients
    /// should retry — this is NOT a permission or invite problem.
    /// Pairs with [`crate::protocol::CLOSE_CODE_TRANSIENT`] so the
    /// close frame actually conveys "retry me".
    pub const STORAGE_ERROR: &str = "STORAGE_ERROR";
    /// The caller's `session_enabled` bit is FALSE — their account
    /// was registered through the public email-signup flow and has
    /// not been approved by an admin yet. The token is still valid
    /// for reads (whoami, history), but session create / attach is
    /// closed. Pairs with the HTTP 403 that `POST /api/sessions`
    /// returns for the same reason. Terminal close, not transient —
    /// the client must route the user to a "pending approval" UI
    /// rather than retry.
    pub const SESSION_DISABLED: &str = "SESSION_DISABLED";
}

/// Terminal WebSocket close code — the client treats this as a
/// permanent refusal and surfaces an error state without reconnecting.
/// Auth / permission / not-found / target-missing all use this.
/// `web/src/lib/ws.ts` switches on `event.code` and MUST stay in sync.
pub const CLOSE_CODE_TERMINAL: u16 = 4001;

/// Transient WebSocket close code — the client is expected to retry
/// (e.g. a one-off storage hiccup during the handshake). Distinct from
/// `CLOSE_CODE_TERMINAL` so the frontend can actually tell the two
/// apart via `event.code`: the preceding `ServerMessage::Error` JSON
/// frame can get dropped if the socket tears down mid-write, so the
/// close code must stand on its own as the retry signal.
///
/// Value sits inside the private-use range (4000-4999) reserved for
/// applications, chosen to be visually distinct from 4001 rather than
/// for any IANA meaning.
pub const CLOSE_CODE_TRANSIENT: u16 = 4503;

/// Map a protocol-level error code to the WebSocket close code that
/// accompanies it. The close code is the ONLY signal the client has
/// for deciding "retry vs give up" — the preceding JSON error frame
/// may be dropped if the socket dies mid-write, so this mapping is
/// the protocol's single source of truth.
///
/// Default is terminal; transient codes must be opted into explicitly
/// so a new error code never silently becomes "silently retry" (which
/// would invite reconnect storms from revoked tokens, etc.).
pub fn close_code_for(error_code: &str) -> u16 {
    match error_code {
        error_codes::STORAGE_ERROR => CLOSE_CODE_TRANSIENT,
        _ => CLOSE_CODE_TERMINAL,
    }
}

/// Stable `reason` codes carried by `ServerMessage::InputDenied`. The
/// frontend switches on these to render localized guidance; keep in sync
/// with `web/src/lib/protocol.ts::InputDeniedReason`.
pub mod input_denied {
    /// Connection's role is `Viewer` — read-only by design.
    pub const VIEWER: &str = "VIEWER";
    /// Session is in `Serialized` mode and caller is not the owner.
    /// Operators can still resize and chat; only keystrokes are dropped.
    pub const SERIALIZED_NOT_OWNER: &str = "SERIALIZED_NOT_OWNER";
}

// --- Client -> Server ---
//
// Terminal input (`TermInput`) is NOT a JSON message — it is sent as a binary
// WebSocket frame carrying raw bytes. The handler at `ws.rs::handle_socket`
// reads `Message::Binary(data)` and forwards it straight to the PTY. JSON
// messages below are used for everything else (session control, resize,
// chat, cursor).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientMessage {
    SessionJoin {
        session_id: String,
        token: String,
        #[serde(default = "default_cols")]
        cols: u16,
        #[serde(default = "default_rows")]
        rows: u16,
    },
    TermResize {
        cols: u16,
        rows: u16,
    },
    ChatMessage {
        text: String,
    },
}

/// Why a participant is being force-disconnected. Carried on
/// [`ServerMessage::PeerEvicted`] so the client can split the UX
/// between "your account was just disabled" (terminal, go to
/// pending-approval) and "your bearer token rotated" (recoverable,
/// go to login). Also shapes the close-frame reason string the
/// server attaches when it tears down the evicted socket.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EvictReason {
    /// Admin flipped `session_enabled = FALSE` (or an equivalent
    /// permanent revoke). Token is dead on every endpoint; the UI
    /// should route the user to the pending-approval page and other
    /// participants should see a "removed by an admin" notice.
    AccountDisabled,
    /// The user's own password change atomically rotated their
    /// bearer token. The session must drop because the old token is
    /// no longer accepted, but the account is in good standing —
    /// the UI should route to login/re-auth and other participants
    /// should see a neutral "re-authentication required" notice,
    /// not an admin action.
    TokenRotated,
}

/// Recording status embedded in `SessionState` so late joiners know
/// immediately whether a recording is active when they connect.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RecordingStatusInfo {
    pub recording_id: String,
    pub started_at: String,
}

// --- Server -> Client ---
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    SessionState {
        session: Session,
        participants: Vec<ParticipantInfo>,
        your_role: Role,
        your_user_id: Uuid,
        /// Bounded replay for late joiners, oldest-first. Ephemeral —
        /// dies with the session; see `session_hub::CHAT_HISTORY_CAP`.
        #[serde(default)]
        chat_history: Vec<ChatEntry>,
        /// Present when a recording is active at join time.
        #[serde(default)]
        recording: Option<RecordingStatusInfo>,
    },
    PeerJoined {
        user_id: Uuid,
        name: String,
        role: Role,
        color: String,
    },
    PeerLeft {
        user_id: Uuid,
    },
    /// Broadcast when a participant is force-removed from a session
    /// *without* them choosing to leave. Distinct from `PeerLeft`
    /// because the meaning for the client UI is different: other
    /// participants render a "was removed" notice and the evicted
    /// user's own WS handler treats this as a close signal (their
    /// session tab must drop immediately so the session can no
    /// longer be driven from their browser).
    ///
    /// `reason` separates the two triggers so the client can split
    /// the UX cleanly: `AccountDisabled` routes the user to the
    /// pending-approval page (their token is now invalid for every
    /// endpoint), while `TokenRotated` routes them to re-login
    /// (their account is fine, only the bearer they were holding is
    /// dead after they rotated their own password). Collaborators
    /// see a different chat string in each case so the same "was
    /// removed" affordance doesn't misattribute a routine password
    /// change as an admin action.
    PeerEvicted {
        user_id: Uuid,
        reason: EvictReason,
    },
    /// Broadcast when the owner changes another participant's role via
    /// `PUT /api/sessions/:id/participants/:user_id/role`. Every
    /// connected client receives this so participant lists update in
    /// lockstep. The WS handler also intercepts messages targeting the
    /// current connection to re-evaluate input permissions without a
    /// reconnect.
    PeerRoleChanged {
        user_id: Uuid,
        new_role: Role,
    },
    PeerChat {
        user_id: Uuid,
        name: String,
        text: String,
        ts: String,
    },
    /// Sent **only to the originating connection** (never broadcast) the
    /// first time a binary input frame is rejected for this session. The
    /// frontend uses this to surface a toast instead of leaving the user
    /// wondering why typing silently does nothing.
    InputDenied {
        /// Machine-readable reason code — see `input_denied` module.
        reason: String,
    },
    Error {
        code: String,
        message: String,
    },
    /// Broadcast to all participants when a recording starts on this session.
    RecordingStarted {
        recording_id: String,
    },
    /// Broadcast to all participants when a recording stops on this session.
    RecordingStopped {
        recording_id: String,
    },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ParticipantInfo {
    pub user_id: Uuid,
    pub name: String,
    pub role: Role,
    pub color: String,
}

/// A single entry in the bounded chat backlog delivered inside
/// `ServerMessage::SessionState` and broadcast live as
/// `ServerMessage::PeerChat`. Fields mirror `PeerChat` 1:1 so the
/// frontend can reuse the same renderer for replayed and live messages
/// — one code path, one formatting story.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ChatEntry {
    pub user_id: Uuid,
    pub name: String,
    pub text: String,
    /// RFC 3339 timestamp captured at broadcast time (not at replay).
    pub ts: String,
}

impl From<ChatEntry> for ServerMessage {
    fn from(e: ChatEntry) -> Self {
        ServerMessage::PeerChat {
            user_id: e.user_id,
            name: e.name,
            text: e.text,
            ts: e.ts,
        }
    }
}

fn default_cols() -> u16 {
    80
}

fn default_rows() -> u16 {
    24
}
